### PR TITLE
Skip thumbnail regeneration when Photon Image CDN handles it externally.

### LINKED ISSUE #30379: Product images with large dimensions break WooCommerce blocks on WordPress.com Atomic sites
**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
When a product using an image with large dimensions is added to a WooCommerce block, the editor crashes with a critical error the next time you try to open it: `There has been a critical error on this website`. 

We have received numerous user reports dating back to around December 2020 and as recently as this week. 

**Expected behavior**
The editor to load successfully.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Open a WordPress.com Atomic site
2. Add a product and upload a product image with large dimensions. (Examples of user images that have triggered this error: 2MB and 9000 x 7000; 40MB and 6000 x 7000. So it seems to be more related to the dimensions rather than the file size.)
3. On any page, add a WooCommerce block, like the Hand-Picked Products block, and then add the test product to the block. 
4. Publish, then try to open the page in the editor again, and see a critical error.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
This issue can be replicated on any WordPress.com Atomic site. I was not able to replicate this on a self-hosted site.

### PR DESCRIPTION
**Pull request by peterfabian**
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #30379  .

tl;dr is that this change skips thumbnail regeneration when Photon Image CDN is active, since it's not needed and it will just take disk space and use CPU cycles unnecessarily, as image requests are redirected to the CDN.

### How to test the changes in this Pull Request:

1. Create a new product, or use an existing one. Go to Product > Set product image
2. Upload an image with large dimensions (size can be modest, e.g. this gradient compresses quite well and it's 16.5k pixels wide)
16_25k
3. On a Photon non-enabled environment (such as local dev env), you will most likely see an error saying that the image could not be processed (as WP attempts to scale down the image and generate thumbnails)
4. Create a business site on wordpress.com (which is Photon enabled), and retry steps 1-3. (If you need help to create a dev site, please see my comment on p9F6qB-9VL-p2). After a couple of attempts, you should be able to upload the image.
5. Try to assign the image to the product
6. Add the Hand picked products block to any post, save it.
7. Try to open the post in WP Admin, it would crash (either the block itself, or the whole site), with an error saying `PHP Fatal error: Allowed memory size of 536870912 bytes exhausted (tried to allocate 61440 bytes) in /srv/htdocs/__wp__/wp-includes/class-wp-image-editor-gd.php on line 111`
8. Apply the patch
9. Test that on Photon non-enabled site, thumbnails have been generated fine. Also test that regenerating thumbnails from WC > Status > Tools works fine (should we switch it to Action Scheduler?)
10. Test that on Photon enabled site, steps 1-7 should work without any problems and it should serve the image via Photon (ie the image comes from `i0.wp.com` and has some parameters at the end, e.g. `?w=1600&ssl=1` or `?fit=1800&ssl=1` or similar, for details, please see https://developer.wordpress.com/docs/photon/)

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/includes/class-wc-regenerate-images.php [hunk 1]
```diff
@@ -466,6 +504,79 @@ public static function queue_image_regeneration() {
 		// Lets dispatch the queue to start processing.
 		self::$background_process->save()->dispatch();
 	}
+
+	/**
+	 * Return the memory limit used for image processing in WP.
+	 *
+	 * @return int Number of bytes of available memory for the PHP process.
+	 */
+	protected static function get_image_memory_limit() {
```
Comment (vedanshujain) [inline plugins/woocommerce/includes/class-wc-regenerate-images.php]: Hey, can we move this and `get_wp_image_memory_limit` function to somewhere in `src` and call from there via a container? that way, we will be able to test and mock them easily.

**Review by vedanshujain — COMMENTED** [review]
Looks good, adding a minor comment, and also this could use some unit tests.

File: plugins/woocommerce/includes/class-wc-regenerate-images.php (hunk 1 above)
Comment (peterfabian) [reply plugins/woocommerce/includes/class-wc-regenerate-images.php]: Thanks, will do that after I'll get the confirmation that functionally it works (ie it stopped crashing).

**Comment by psealock** [comment]
@peterfabian Is progress still happening on this? If not can we close it?

**Comment by peterfabian** [comment]
I need to add some unit tests to it, then I was hoping we can merge it, but it got deprioritized amidst other stuff. Will try to get it done in a week or so. Thanks for the nudge.
